File: Code/GeneSimulation_py/Server/Bots/chromosome.py
class Chromosome:
    def __init__(self, chromosome):
        self.chromosome = chromosome
        self.fitness = 0

    def set_fitness(self, fitness):
        self.fitness = fitness

    def add_fitness(self, new_fitness):
        self.fitness += new_fitness

    def reset_fitness(self):
        self.fitness = 0

    # Override __getitem__ to allow direct access to chromosome list
    def __getitem__(self, index):
        return self.chromosome[index]

    # Override __eq__
    def __eq__(self, other):
        return isinstance(other, Chromosome) and self.chromosome == other.chromosome

    def __hash__(self):
        return hash(tuple(self.chromosome))

File: Code/GeneSimulation_py/cppSocialChoiceGenes/GeneticStuff/Chromosome.cpp
//
// Created by Sean on 4/8/2025.
//

#include "Chromosome.h"
#include <vector>


// Constructor
Chromosome::Chromosome(const std::vector<double>& chromosome)
    : chromosome(chromosome), fitness(0) {}

std::vector<double> Chromosome::getChromosome() const {
    return chromosome;
}

// so thats how we do that. this->Fitness (our fitness class object) gets set to fitness
void Chromosome::setFitness(double fitness) {
    this->fitness = fitness;
}

// Add to the fitness value
void Chromosome::addFitness(double newFitness) {
    this->fitness += newFitness;
}

// Reset the fitness value
void Chromosome::resetFitness() {
    this->fitness = 0;
}

int Chromosome::getFitness() const {
    return this->fitness;
}

// overloaidng the [] for direct access
double Chromosome::operator[](size_t index) const {
    return chromosome[index];
}

// Overload the == operator to compare chromosomes
bool Chromosome::operator==(const Chromosome& other) const {
    return this->chromosome == other.chromosome;
}

// custom hash function so we can use it to store the chromosome as a key in a dict.
size_t Chromosome::ChromosomeHash::operator()(const Chromosome& c) const {
    // Hash the chromosome as a vector of doubles
    size_t seed = 0;
    for (const auto& value : c.chromosome) {
        seed ^= std::hash<double>{}(value) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
    }
    return seed;
}